File: archive/python-legacy/polymarket/data_pipeline/tomorrow_io_fetcher.py
"""
Tomorrow.io weather data fetcher module
Retrieves historical weather data from Tomorrow.io Timeline API
Requires API key, global coverage
"""

import logging
import requests
import pandas as pd
from datetime import datetime, timedelta
from typing import List, Optional

from polymarket.config import TOMORROW_IO_API_KEY

logger = logging.getLogger(__name__)

LOCATIONS = {
    "NYC": {"lat": 40.71, "lon": -74.01, "name": "New York, NY"},
    "LA": {"lat": 34.05, "lon": -118.24, "name": "Los Angeles, CA"},
    "London": {"lat": 51.51, "lon": -0.13, "name": "London, UK"},
    "Chicago": {"lat": 41.88, "lon": -87.63, "name": "Chicago, IL"},
    "Dallas": {"lat": 32.78, "lon": -96.80, "name": "Dallas, TX"},
    "Denver": {"lat": 39.74, "lon": -104.99, "name": "Denver, CO"},
    "Miami": {"lat": 25.76, "lon": -80.19, "name": "Miami, FL"},
    "Boston": {"lat": 42.36, "lon": -71.06, "name": "Boston, MA"},
}

BASE_URL = "https://api.tomorrow.io/v4/timelines"


class TomorrowIOFetcher:
    """Fetches historical weather data from Tomorrow.io API"""

    def __init__(self, api_key: str = TOMORROW_IO_API_KEY):
        self.api_key = api_key
        self.session = requests.Session()

    def is_available(self) -> bool:
        """Check if API key is configured"""
        return bool(self.api_key)

    def fetch_daily_observations(
        self,
        latitude: float,
        longitude: float,
        start_date: str,
        end_date: str,
        location_id: str = "TIO",
    ) -> pd.DataFrame:
        """
        Fetch daily weather observations for a location

        Args:
            latitude: Location latitude
            longitude: Location longitude
            start_date: Start date (YYYY-MM-DD)
            end_date: End date (YYYY-MM-DD)
            location_id: Identifier for the location

        Returns:
            DataFrame with standardized weather observations
        """
        if not self.is_available():
            logger.info("Tomorrow.io: No API key configured, skipping")
            return pd.DataFrame()

        params = {
            "location": f"{latitude},{longitude}",
            "fields": ",".join([
                "temperatureMax",
                "temperatureMin",
                "temperatureAvg",
                "precipitationIntensityAvg",
                "windSpeedAvg",
            ]),
            "timesteps": "1d",
            "startTime": f"{start_date}T00:00:00Z",
            "endTime": f"{end_date}T23:59:59Z",
            "apikey": self.api_key,
            "units": "metric",
        }

        try:
            response = self.session.get(BASE_URL, params=params, timeout=30)
            response.raise_for_status()
            data = response.json()

            timelines = data.get("data", {}).get("timelines", [])
            if not timelines:
                logger.warning(f"Tomorrow.io: No data for {location_id}")
                return pd.DataFrame()

            observations = []
            for timeline in timelines:
                for interval in timeline.get("intervals", []):
                    values = interval.get("values", {})
                    start_time = interval.get("startTime", "")

                    observations.append({
                        "date": pd.Timestamp(start_time[:10]),
                        "station_id": f"TIO_{location_id}",
                        "temperature_max": values.get("temperatureMax"),
                        "temperature_min": values.get("temperatureMin"),
                        "temperature_mean": values.get("temperatureAvg"),
                        "precipitation_total": (
                            values.get("precipitationIntensityAvg", 0) * 24
                        ),  # mm/hr to mm/day
                        "wind_speed_mean": values.get("windSpeedAvg"),
                    })

            if not observations:
                return pd.DataFrame()

            df = pd.DataFrame(observations)
            logger.info(f"Tomorrow.io: Fetched {len(df)} observations for {location_id}")
            return df

        except requests.RequestException as e:
            logger.error(f"Tomorrow.io error for {location_id}: {e}")
            return pd.DataFrame()

    def fetch_location(
        self,
        location_key: str,
        start_date: str,
        end_date: str,
    ) -> pd.DataFrame:
        """Fetch data for a pre-configured location"""
        location = LOCATIONS.get(location_key)
        if not location:
            logger.error(f"Unknown location: {location_key}")
            return pd.DataFrame()

        return self.fetch_daily_observations(
            latitude=location["lat"],
            longitude=location["lon"],
            start_date=start_date,
            end_date=end_date,
            location_id=location_key,
        )

    def fetch_multiple_locations(
        self,
        location_keys: List[str],
        start_date: str,
        end_date: str,
    ) -> pd.DataFrame:
        """Fetch data for multiple locations and combine"""
        all_data = []
        for key in location_keys:
            df = self.fetch_location(key, start_date, end_date)
            if not df.empty:
                all_data.append(df)

        if all_data:
            return pd.concat(all_data, ignore_index=True)
        return pd.DataFrame()

File: src/data_pipeline/tomorrow_io_fetcher.rs
use std::collections::HashMap;

use chrono::NaiveDate;
use serde_json::Value;

use crate::config::tomorrow_io_api_key;
use crate::types::WeatherRecord;

const BASE_URL: &str = "https://api.tomorrow.io/v4/timelines";

#[derive(Clone)]
pub struct TomorrowIOFetcher {
    api_key: String,
    client: reqwest::blocking::Client,
    locations: HashMap<String, (f64, f64)>,
}

impl TomorrowIOFetcher {
    pub fn new(api_key: Option<String>) -> Self {
        Self {
            api_key: api_key.unwrap_or_else(tomorrow_io_api_key),
            client: crate::data_pipeline::build_client(30),
            locations: default_locations(),
        }
    }

    pub fn is_available(&self) -> bool {
        !self.api_key.is_empty()
    }

    pub fn fetch_daily_observations(
        &self,
        latitude: f64,
        longitude: f64,
        start_date: NaiveDate,
        end_date: NaiveDate,
        location_id: &str,
    ) -> Vec<WeatherRecord> {
        if !self.is_available() {
            return Vec::new();
        }

        let params = [
            ("location", format!("{latitude},{longitude}")),
            (
                "fields",
                "temperatureMax,temperatureMin,temperatureAvg,precipitationIntensityAvg,windSpeedAvg".to_string(),
            ),
            ("timesteps", "1d".to_string()),
            (
                "startTime",
                format!("{}T00:00:00Z", start_date.format("%Y-%m-%d")),
            ),
            (
                "endTime",
                format!("{}T23:59:59Z", end_date.format("%Y-%m-%d")),
            ),
            ("apikey", self.api_key.clone()),
            ("units", "metric".to_string()),
        ];

        let Ok(resp) = self.client.get(BASE_URL).query(&params).send() else {
            return Vec::new();
        };
        let Ok(json) = resp.json::<Value>() else {
            return Vec::new();
        };

        let intervals = json
            .get("data")
            .and_then(|d| d.get("timelines"))
            .and_then(|v| v.as_array())
            .and_then(|arr| arr.first())
            .and_then(|t| t.get("intervals"))
            .and_then(|v| v.as_array())
            .cloned()
            .unwrap_or_default();

        let mut out = Vec::new();
        for interval in intervals {
            let start_time = interval.get("startTime").and_then(|v| v.as_str());
            let Some(start_time) = start_time else {
                continue;
            };

            let Ok(date) = NaiveDate::parse_from_str(&start_time[..10], "%Y-%m-%d") else {
                continue;
            };
            let values = interval.get("values").cloned().unwrap_or(Value::Null);

            let mut row = WeatherRecord::new(date, format!("TIO_{location_id}"));
            row.temperature_max = values.get("temperatureMax").and_then(|v| v.as_f64());
            row.temperature_min = values.get("temperatureMin").and_then(|v| v.as_f64());
            row.temperature_mean = values.get("temperatureAvg").and_then(|v| v.as_f64());
            row.precipitation_total = values
                .get("precipitationIntensityAvg")
                .and_then(|v| v.as_f64())
                .map(|x| x * 24.0);
            row.wind_speed_mean = values.get("windSpeedAvg").and_then(|v| v.as_f64());
            out.push(row);
        }

        out
    }

    pub fn fetch_location(&self, location_key: &str, start_date: NaiveDate, end_date: NaiveDate) -> Vec<WeatherRecord> {
        let Some((lat, lon)) = self.locations.get(location_key) else {
            return Vec::new();
        };
        self.fetch_daily_observations(*lat, *lon, start_date, end_date, location_key)
    }
}

impl Default for TomorrowIOFetcher {
    fn default() -> Self {
        Self::new(None)
    }
}

fn default_locations() -> HashMap<String, (f64, f64)> {
    [
        ("NYC", (40.71, -74.01)),
        ("LA", (34.05, -118.24)),
        ("London", (51.51, -0.13)),
        ("Chicago", (41.88, -87.63)),
        ("Dallas", (32.78, -96.80)),
        ("Denver", (39.74, -104.99)),
        ("Miami", (25.76, -80.19)),
        ("Boston", (42.36, -71.06)),
    ]
    .into_iter()
    .map(|(k, v)| (k.to_string(), v))
    .collect()
}
